amethyst_rendy/src/submodules/vertex.rs

```diff
@@ -11,25 +14,121 @@ use crate::{
-pub struct DynamicVertex<B: Backend, T: 'static> {
-    per_image: Vec<PerImageDynamicVertex<B>>,
-    marker: PhantomData<T>,
+pub struct IndexData<T>(PhantomData<T>);
+
+/// Type used to compile-time specify the type of vertex buffer data managed by a  `DynamicVertexData`
+#[derive(Debug)]
+pub struct VertexData<T>(PhantomData<T>);
+
+/// Type trait for allowing type-based implementation details for binding the different buffer types
+/// of index and vertex `DynamicVertexData`
+pub trait VertexDataBufferType<B: Backend> {
+    /// Returns this type implementations `gfx_hal::buffer::Usage`
+    fn usage() -> hal::buffer::Usage;
+
+    /// Executes this types specific binding implementation
+    fn bind(
+        binding_id: u32,
+        encoder: &mut RenderPassEncoder<'_, B>,
+        buffer: &Option<Escape<Buffer<B>>>,
+        offset: u64,
+    ) -> bool;
+}
+
+impl<B: Backend, T: 'static> VertexDataBufferType<B> for VertexData<T> {
+    #[inline]
+    fn usage() -> hal::buffer::Usage {
+        hal::buffer::Usage::VERTEX
+    }
+
+    #[inline]
+    fn bind(
+        binding_id: u32,
+        encoder: &mut RenderPassEncoder<'_, B>,
+        buffer: &Option<Escape<Buffer<B>>>,
+        offset: u64,
+    ) -> bool {
+        if let Some(buffer) = buffer.as_ref() {
+            encoder.bind_vertex_buffers(binding_id, Some((buffer.raw(), offset)));
+            return true;
+        }
+
+        false
+    }
+}
+
+impl<B: Backend> VertexDataBufferType<B> for IndexData<u16> {
+    #[inline]
+    fn usage() -> hal::buffer::Usage {
+        hal::buffer::Usage::INDEX
+    }
+
+    #[inline]
+    fn bind(
+        _: u32,
+        encoder: &mut RenderPassEncoder<'_, B>,
+        buffer: &Option<Escape<Buffer<B>>>,
+        offset: u64,
+    ) -> bool {
+        if let Some(buffer) = buffer.as_ref() {
+            encoder.bind_index_buffer(buffer.raw(), offset, hal::IndexType::U16);
+            return true;
+        }
+
+        false
+    }
 }
 
+impl<B: Backend> VertexDataBufferType<B> for IndexData<u32> {
+    #[inline]
+    fn usage() -> hal::buffer::Usage {
+        hal::buffer::Usage::INDEX
+    }
+
+    #[inline]
+    fn bind(
+        _: u32,
+        encoder: &mut RenderPassEncoder<'_, B>,
+        buffer: &Option<Escape<Buffer<B>>>,
+        offset: u64,
+    ) -> bool {
+        if let Some(buffer) = buffer.as_ref() {
+            encoder.bind_index_buffer(buffer.raw(), offset, hal::IndexType::U32);
+            return true;
+        }
+
+        false
+    }
+}
+
+/// This structure wraps [PerImageDynamicVertexData], managing multiple instances and providing
+/// an easy-to-use interface for having per-image buffers. This is needed because multiple images
+/// (frames) can be in flight at any given time, so multiple buffers are needed for the same data.
 #[derive(Debug)]
-struct PerImageDynamicVertex<B: Backend> {
-    buffer: Option<Escape<Buffer<B>>>,
+pub struct DynamicVertexData<B: Backend, V: VertexDataBufferType<B>, T: 'static> {
```

Comment:
i believe you don't need `T` for anything here.

Reply:
It's used below at `I::Item: AsRef<[T]>,`